hybrasyl/CreatureStatus.cs

```diff
@@ -136,39 +151,38 @@ public class CreatureStatus : ICreatureStatus
 
         public double ElapsedSinceTick => (DateTime.Now - LastTick).TotalSeconds;
 
-        private void _init(Status xmlstatus, Creature target, Castable castable)
+        public CreatureStatus(Status xmlstatus, Creature target, Creature source=null, Castable castable = null,
+            int durationOverride = -1, int tickOverride = -1)
         {
-            XMLStatus = xmlstatus;
             Target = target;
+            Source = source;
+            XMLStatus = xmlstatus;
             Castable = castable;
-
-        }
-        public CreatureStatus(Status xmlstatus, Creature target, Castable castable = null,
-            int durationOverride = -1, int tickOverride = -1)
-        {
-            _init(xmlstatus, target, castable);
             Start = DateTime.Now;
             _durationOverride = durationOverride;
             _tickOverride = tickOverride;
         }
 
-        public CreatureStatus(Status xmlstatus, Creature target, Castable castable = null)
+        public CreatureStatus(Status xmlstatus, Creature target, Creature source=null, Castable castable = null)
         {
-            _init(xmlstatus, target, castable);
+            Target = target;
+            Source = source;
+            XMLStatus = xmlstatus;
+            Castable = castable;
             Start = DateTime.Now;
 
-            var addList = castable?.Statuses.Add.Where(e => e.Value == xmlstatus.Name);
+            var addList = castable?.Effects.Statuses.Add.Where(e => e.Value == xmlstatus.Name);
             if (addList?.Count() > 0)
             {
                 var addObj = addList.First();
-                _durationOverride = addObj.Duration * xmlstatus.Duration;
+                _durationOverride = addObj.Duration != 0 ? addObj.Duration : xmlstatus.Duration;
                 _tickOverride = (int) Math.Floor(addObj.Speed * xmlstatus.Tick);
             }
         }
 
         private void ProcessSfx(ModifierEffect effect)
         {
-            if (effect.Sound?.Id != null)
+            if (effect.Sound?.Id != 0)
```

Comment:
Sound is nullable. A null check on Id should apply.

Reply:
Sound is not nullable as exists currently. This check was failing, which is why it was changed. `Sound` always has a default value.

Reply:
Looking at that block, sound is indeed nullable.

Reply:
It is not a nullable type in the structure `xsd2code` generates, is what I mean. My expectation that it is by use of `?` is wrong. `Status.Designer.cs` line 1210; also note it is always initialized in `ModifierEffect` (line 1116).

If you have a better suggestion for how to handle this case we see elsewhere, I welcome it. We may need to make more fields nullable as it is untenable for many things like this to have a default of `0` as `0` is in fact a valid sound. This fix in particular though is a workaround to the `weird sound playing every status tick` bug because previously this is never null and will always result in a sound playing. Cast `Armachd` and observe the results with the null check in place.